DM-30140: Allow trash to take iterable of dataset refs

timj opened the pull request:
This is much much faster than calling trash on each ref separately. Also remove the existence check since if we are asking for a dataset to be trashed it is not really that important if the dataset has already disappeared and that could just be caused by someone else pruning a collection in parallel.

timj committed:
Allow trash to take iterable of dataset refs

This is much much faster than calling trash on each ref
separately. Also remove the existence check since if we are
asking for a dataset to be trashed it is not really that
important if the dataset has already disappeared and that
could just be caused by someone else prunning a collection
in parallel.

TallJimbo commented on python/lsst/daf/butler/datastores/genericDatastore.py:
```diff
@@ -148,7 +148,7 @@ def _move_to_trash_in_registry(self, ref: DatasetRef) -> None:
         # have been deleted already from registry but are still in
         # the python object. moveToTrash will deal with that.
-        self.bridge.moveToTrash([ref])
+        self.bridge.moveToTrash(refs)
```
Might be time to just remove `_move_to_trash_in_registry` and call this method directly.

TallJimbo commented on python/lsst/daf/butler/datastores/chainedDatastore.py:
```diff
@@ -629,8 +629,13 @@ def forget(self, refs: Iterable[DatasetRef]) -> None:
+        if isinstance(ref, DatasetRef):
+            ref_label = str(ref)
+        else:
+            ref_label = "bulk datasets"
```
Might be nice to put `len(ref)` in here.

timj replied:
Wouldn't that be a problem if ref was a generator? I assumed that `len()` would trigger the generator and empty it for later.

TallJimbo replied:
Oh, good point.  If it's a generator or other lazy iterator `len` wouldn't work at all. Nevermind.

timj committed:
Remove _move_to_trash_in_registry and call bridge directly.

The indirection was not helping anything since it only included
one call that was generic.

timj committed:
Update copyright dates
